### PR TITLE
POC: optional level-3 TOC

### PR DESCRIPTION
**Pull request by osfameron**
This commit adds the `[.include-in-toc]` role which will select h3 to be
included in the table of contents.

Basic CSS is provided to differentiate h1, h2, h3.

(No idea if this is the best / a half-reasonable way of doing this, asking @couchbase/docs-writers for review!)

### CODE REVIEW

**Comment by osfameron** [comment]
Screenshot 2021-11-02 at 18 04 10

Sample screenshot. The h1 is now bold, and the h3 have a list-indicator.

**Comment by osfameron** [comment]
Prompted by discussion on https://github.com/couchbase/docs-sdk-java/pull/289

File: src/css/toc.css [hunk 1]
```diff
@@ -112,3 +112,19 @@
...
   color: inherit;
 }
+
+.toc .toc-menu li.toc-H1 {
+  font-weight: bold;
+  font-size: 110%;
+}
+
+.toc .toc-menu li.toc-H3 {
+  list-style-type: circle;
```
Comment (simon-dew) [inline src/css/toc.css]: Just a personal preference, but I don't really like the circle list style in the toc. Would you object to a hyphen?

```suggestion
  list-style-type: '-';
```

**Review by simon-dew — CHANGES_REQUESTED** [review]
@osfameron this is a great idea and I can see it would be useful. I'm just wondering whether it would be better to use markup that uses the standard AsciiDoctor attributes? https://docs.asciidoctor.org/asciidoc/latest/toc/section-depth/

If writers specify `:toclevels: 3` and want to _exclude_ a particular level 3 heading from the toc, the standard method is then to apply the `[discrete]` style to that level 3 heading.

File: src/css/toc.css (hunk 1 above)
Comment (simon-dew) [reply src/css/toc.css]: Also not sure why you've made heading 1 bold. There should only be one heading 1 in the page (the page title) and it shouldn't appear in the table of contents! (Some of our pages do contain more than one heading 1, but this is something we should be fixing.)

**Comment by simon-dew** [comment]
> I'm just wondering whether it would be better to use markup that uses the standard AsciiDoctor attributes?

Since this is a standard AsciiDoctor feature, Antora might well support it already, so we might just need to switch the feature back on for our implementation. Maybe worth talking to @RayOffiah — he knows a lot about this sort of thing 😉

**Comment by RayOffiah** [comment]
@osfameron 

Can't we just use toclevel?

https://docs.asciidoctor.org/asciidoc/latest/toc/section-depth/

If you put this in the playbook:

```
asciidoc:
  attributes:
    page-toclevels: 2@
```

This will default all the tocs to level 2, but you can override on individual pages by adding the toclevel attribute to the page

File: src/css/toc.css (hunk 1 above)
Comment (osfameron) [reply src/css/toc.css]: Yeah, that was just as the demo page on preview does have a h1, can remove.

**Comment by osfameron** [comment]
@RayOffiah @simon-dew thanks, I'll have a look at toclevels!

It looked to me like this was controlled in src/js/02-on-this-page.js but will have a play.

**Comment by osfameron** [comment]
I couldn't get the :toclevels: attributes to work, and don't find evidence of them being used by our current UI.
I've logged https://support.opendevise.com/support/tickets/115 to ask Dan & co to advise!

**Comment by osfameron** [comment]
@simon-dew @RayOffiah re-requesting review.

Omissions:

 * not implemented `[discrete]` handling as it wasn't in the bundle from Opendevise, and I'm not sure if it's required yet?
 * are there pages that I need to test that for, to make sure there's no regression?
 * whole thing will require some testing (also because couldn't figure out how to set asciidoc *defaults* in docs-ui preview and need to test that locally. In the meantime, I've defaulted the toclevels to *1* rather than 2 as per Antora default UI, to avoid making massive changes)

**Comment by simon-dew** [comment]
Thanks @osfameron ! Just one question to start with — why `page-toclevels` instead of just `toclevels`? I'd like to stick to standard AsciiDoctor syntax as much as possible. I may be missing some context here because I can't view the OpenDevise ticket.

**Comment by simon-dew** [comment]
> not implemented `[discrete]` handling as it wasn't in the bundle from Opendevise, and I'm not sure if it's required yet?

You'll need it if you only want _some_ of the lower-level headings to appear in the ToC, and not all of them.

**Comment by simon-dew** [comment]
> In the meantime, I've defaulted the toclevels to _1_ rather than 2 as per Antora default UI, to avoid making massive changes

So just to clarify, the default is that the ToC will only pick up heading 2, as at present?

**Comment by osfameron** [comment]
> You'll need it if you only want _some_ of the lower-level headings to appear in the ToC, and not all of them.

Yes, I'll test and implement if I require it, OK to omit if it isn't immediately needed though presumably?

**Comment by osfameron** [comment]
> So just to clarify, the default is that the ToC will only pick up heading 2, as at present?

That's the intention! I'll test for regressions locally and then again on staging when we get there...

**Comment by simon-dew** [comment]
> > You'll need it if you only want _some_ of the lower-level headings to appear in the ToC, and not all of them.
> 
> Yes, I'll test and implement if I require it, OK to omit if it isn't immediately needed though presumably?

Sure, I was just thinking of your original use-case, in which you were going to specify individual heading 3s 😄

**Comment by osfameron** [comment]
Looks like `[discrete]` *is* in fact supported. Not quite sure how, or why my test didn't demonstrate this in docs-ui preview, apologies for confusion.

File: gulp.d/tasks/build-preview-pages.js [hunk 2]
```diff
@@ -116,6 +121,9 @@ module.exports =
                     pageModel.attributes = Object.entries({ ...attributes, ...componentVersion.asciidoc.attributes })
                       .filter(([name, val]) => name.startsWith('page-'))
                       .reduce((accum, [name, val]) => ({ ...accum, [name.substr(5)]: val }), {})
+
+                    // console.log(pageModel.attributes)
```
Comment (maria-robobug) [inline gulp.d/tasks/build-preview-pages.js]: 🚗 Drive-by comment...
Guessing this is for testing things out, just a reminder to remove once you're happy with it. 📆

**Comment by simon-dew** [comment]
I see that `page-toclevels` is standard for Antora. LGTM 👍

**Review by simon-dew — APPROVED** [review]

**Review by maria-robobug — APPROVED** [review]
👍 Don't forget to remove commented out `console.log` in https://github.com/couchbase/docs-ui/pull/126/files#diff-c1e8fc5ee1fed60ec0eb78bd527b66e476e6f83f10f83cb2bcecc5f7b0be149dR31

**Pull request merged**